libgalois/src/analytics/matrix_completion/matrix_completion.cpp

```diff
@@ -0,0 +1,534 @@
+
+  katana::StatTimer executeAlgoTimer("Algorithm Execution Time");
+  katana::TimeAccumulator elapsed;
+  elapsed.start();
+
+  unsigned long lastTime = 0;
+
+  for (unsigned int round = 0;; round += deltaRound) {
+    if (plan.fixedRounds() > 0 && round >= plan.fixedRounds())
+      break;
+    if (plan.fixedRounds() > 0)
+      deltaRound = std::min(deltaRound, plan.fixedRounds() - round);
+
+    for (unsigned i = 0; i < plan.updatesPerEdge(); ++i) {
+      // Assume that loss decreases
+      if (sf.isBold())
+        steps[i] = i == 0 ? rate : steps[i - 1] * 1.05;
+      else
+        steps[i] = sf.stepSize(round + i, plan);
+    }
+
+    executeAlgoTimer.start();
+    fn(&steps[0], round + deltaRound, plan.useExactError() ? &errorAccum : NULL, plan);
+    executeAlgoTimer.stop();
+    double error = plan.useExactError() ? errorAccum.reduce() : sumSquaredError(graph);
+
+    elapsed.stop();
+
+    unsigned long curElapsed = elapsed.get();
+    elapsed.start();
+    unsigned long millis = curElapsed - lastTime;
+    lastTime = curElapsed;
+
+    double gflops = countFlops(graph.num_edges(), deltaRound, LATENT_VECTOR_SIZE, plan) /
+                    millis / 1e6;
+
+    int curRound = round + deltaRound;
+    katana::gPrint(
+        "R: ", curRound, " elapsed (ms): ", curElapsed, " GFLOP/s: ", gflops);
+    if (plan.useExactError()) {
+      katana::gPrint(
+          " RMSE (R ", curRound, "): ", std::sqrt(error / graph.num_edges()), "\n");
+    } else {
+      katana::gPrint(
+          " Approx. RMSE (R ", (curRound - 1),
+          ".5): ", std::sqrt(std::abs(error / graph.num_edges())), "\n");
+    }
+
+    katana::gPrint("Error Change : ", std::abs((last - error) / last), "\n");
+    if (!isFinite(error))
+      break;
+    if (plan.fixedRounds() <= 0 &&
+        (round >= plan.maxUpdates() || std::abs((last - error) / last) < plan.tolerance()))
+      break;
+    if (sf.isBold()) {
+      // Assume that loss decreases first round
+      if (last >= 0.0 && last < error)
+        rate = steps[deltaRound - 1] * 0.5;
+      else
+        rate = steps[deltaRound - 1] * 1.05;
+    }
+    last = error;
+  }
+}
+
+
+class SGDItemsAlgo {
+
+public:
+  bool isSgd() const { return true; }
+
+  std::string name() const { return "sgdItemsAlgo"; }
+
+  size_t numItems() const { return NUM_ITEM_NODES; }
+
+private:
+  struct Execute {
+    Graph& graph;
+    katana::GAccumulator<unsigned>& edges_visited;
+
+    void operator()(
+        LatentValue* steps, int, katana::GAccumulator<double>* errorAccum, MatrixCompletionPlan plan) {
+      const LatentValue step_size = steps[0];
+      katana::do_all(
+          katana::iterate(graph.begin(), graph.begin() + NUM_ITEM_NODES), [&](GNode src) {
+            for (auto ii : graph.edges(src)) {
+              auto dst = graph.GetEdgeDest(ii);
+              auto& item_latent_vector = graph.GetData<NodeLatentVector>(src);
+              auto& user_latent_vector = graph.GetData<NodeLatentVector>(dst);
```

Comment:
GetEdgeDest should return a pointer.  I think you need to use "*dst", isn't it?

Reply:
You are right, GetEdgeDest returns a node_iterator, but I think there is an overloaded version of GetData with node_iterator as a parameter.